README.md
![GitHub Actions](https://github.com/mmazzarolo/serverino/workflows/Lint%20&%20Test/badge.svg)
[![Install Size](https://packagephobia.now.sh/badge?p=serverino)](https://packagephobia.now.sh/result?p=serverino)

<br />

# Serverino! <img src="./.github/banner.png" width="110" align="left">

Tiny CLI-based static server

<br />
<br />

Serverino is a lightweight tool for serving static content on localhost.

Thanks to locally-trusted development certificates it can also host your static content on https://localhost.

No setup required.

Serverino is inspired by [Zeit's Serve](https://github.com/zeit/serve) and is powered by [Express](https://expressjs.com/) and [https-localhost](https://github.com/daquinoaldo/https-localhost).

## Usage

The quickest way to get started is to run serverino in your project directory using [npx](https://blog.npmjs.org/post/162869356040/introducing-npx-an-npm-package-runner).

```
npx serverino
```

If you prefer, you can also install the package globally using [npm](https://npmjs.com/) or [Yarn](https://yarnpkg.com/).

```bash
# Using npm
npm install --global serverino

# Using yarn
yarn global add serverino
```

Once that's done, you can run this command inside your project's directory...

```bash
serverino
```

...or specify which folder you want to serve:

```bash
serverino folder_name
```

That's it! :tada:

## Available options

```
Usage
    $ serverino [dir] [path]

  Options
    --port -p     Port
    --secure -s   Use HTTPS
    --cors -c     Set "Access-Control-Allow-Origin" to "*"
    --verbose -v  Log requests

  Examples
    $ serverino
    $ serverino ./dist
    $ serverino ./dist /admin/

  Run without arguments serves the current directory on http:/localhost:8080/.
```

## Contributing

Each contribute is welcome!  
Please, checkout the [contributing guidelines](./CONTRIBUTING.md).

## License

Is released under [AGPL-3.0 - GNU Affero General Public License v3.0](./LICENSE.md).

### Briefly:

- modification and redistribution allowed for both private and **commercial use**
- you must **grant patent rigth to the owner and to all the contributors**
- you must **keep it open source** and distribute under the **same license**
- changes must be documented
- include a limitation of liability and it **does not provide any warranty**

### Warranty

THIS TOOL IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND.
THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE PROGRAM IS WITH YOU.
For the full warranty check the [LICENSE](./LICENSE.md).


CHANGELOG.md
# [1.1.0](https://github.com/mmazzarolo/serverino/compare/v1.0.2...v1.1.0) (2020-08-29)


### Features

* Automatically add leading slash to the path ([#8](https://github.com/mmazzarolo/serverino/issues/8)) ([7327c28](https://github.com/mmazzarolo/serverino/commit/7327c281a8d4cdf0c155d2fe3e1f8b035432e5ce))

## [1.0.2](https://github.com/mmazzarolo/serverino/compare/v1.0.1...v1.0.2) (2020-07-22)


### Bug Fixes

* **security:** Dependencies update ([#7](https://github.com/mmazzarolo/serverino/issues/7)) ([90f1e57](https://github.com/mmazzarolo/serverino/commit/90f1e579c14fecc2cabb7f25bc4015ffa82d36d3))

## [1.0.1](https://github.com/mmazzarolo/serverino/compare/v1.0.0...v1.0.1) (2020-05-06)


### Bug Fixes

* Testing fix bump ([0623bed](https://github.com/mmazzarolo/serverino/commit/0623bed68d18df398bb69647515ed8e627c872d6))

# 1.0.0 (2020-05-06)


### Bug Fixes

* --help command output improvements ([#4](https://github.com/mmazzarolo/serverino/issues/4)) ([840700a](https://github.com/mmazzarolo/serverino/commit/840700ae58fea31dc7eb322712b1e97bb58d135d))


index.test.js
/* eslint-disable jest/expect-expect */
const request = require("supertest");
const net = require("net");

const createApp = require("./index.js");
const consoleInfoSpy = jest.spyOn(console, "info");

beforeEach(function () {
  global.console.info.mockImplementation(() => {});
});

afterEach(function () {
  jest.clearAllMocks();
});

test("serves from current path", async function () {
  const app = createApp();
  await request(app.serve(false)).get("/package.json").expect(200);
  await request(app.serve(false)).get("/i-do-not-exist.json").expect(404);
});

test("serves on a custom path", async function () {
  const app = createApp({
    path: "/test",
  });
  await request(app.serve(false)).get("/test/package.json").expect(200);
  await request(app.serve(false)).get("/test/i-do-not-exist.json").expect(404);
});

test.todo("serves on SSL");

test("serves on a custom port", async function () {
  const isPortTaken = (port) =>
    new Promise((resolve, reject) => {
      const tester = net
        .createServer()
        .once("error", (err) =>
          err.code == "EADDRINUSE" ? resolve(true) : reject(err)
        )
        .once("listening", () =>
          tester.once("close", () => resolve(false)).close()
        )
        .listen(port);
    });
  const port = 1337;
  expect(await isPortTaken(port)).toBeFalsy();
  const app = createApp({
    port: port,
  }).serve();
  expect(await isPortTaken(port)).toBeTruthy();
  app.close();
});

test(`serves with "Access-Control-Allow-Origin" set to "*"`, async function () {
  const app = createApp({
    cors: true,
  });
  await request(app.serve(false))
    .get("/package.json")
    .expect(200)
    .expect("Access-Control-Allow-Origin", "*");
});

test(`serves verbosely`, async function () {
  const app = createApp({
    verbose: true,
  });
  await request(app.serve(false)).get("/package.json").expect(200);
  expect(consoleInfoSpy).toHaveBeenCalledWith(
    expect.stringMatching(/\/package.json/)
  );
});


index.js
#!/usr/bin/env node
"use strict";

const express = require("express");
const resolve = require("path").resolve;
const https = require("https");
const certs = require("https-localhost/certs");
const chalk = require("chalk");

const createApp = function (options = {}) {
  const {
    root = process.cwd(),
    path = "/",
    domain = "localhost",
    verbose = false,
    port = process.env.PORT || 8080,
    cors = false,
    secure = false,
  } = options;

  const app = express();

  if (secure) {
    app.listen = async function (...args) {
      app.server = https
        .createServer(await certs.getCerts(domain), app)
        .listen(...args);
      return app.server;
    };
  }

  if (cors) {
    app.use(function (req, res, next) {
      res.header("Access-Control-Allow-Origin", "*");
      next();
    });
  }

  if (verbose) {
    app.use(function (req, res, next) {
      const time = new Date().toISOString().substr(11, 8);
      console.info(
        chalk.dim(`[${time}] › `) + chalk.green(`${req.method} `) + req.url
      );
      next();
    });
  }

  app.serve = function (shouldStartListening = true) {
    app.use(path, express.static(resolve(root)));
    if (shouldStartListening) {
      return app.listen(port, function () {
        const servedDir = options.root ? options.root : "the current directory";
        const address = `http${secure ? "s" : ""}://localhost:${port}${
          path ? path : ""
        }`;
        console.info(`Serving ${servedDir} on ${chalk.yellowBright(address)}`);
      });
    }
    return app;
  };

  return app;
};

module.exports = createApp;

if (require.main === module) {
  createApp().serve();
}


cli.js
#!/usr/bin/env node
"use strict";

const meow = require("meow");
const serverino = require("./index.js");

const cli = meow(
  `
  Usage
    $ serverino [dir] [path]
  Options
    --port -p     Port
    --secure -s   Use HTTPS
    --cors -c     Set "Access-Control-Allow-Origin" to "*"
    --verbose -v  Log requests
  Examples
    $ serverino
    $ serverino ~/build
    $ serverino dist /admin/
  Run without arguments serves the current directory on http:/localhost:8080/.
`,
  {
    inferType: true,
    flags: {
      port: {
        type: "number",
        alias: "p",
        default: 8080,
      },
      secure: {
        type: "boolean",
        alias: "s",
        default: false,
      },
      cors: {
        type: "boolean",
        alias: "c",
        default: false,
      },
      verbose: {
        type: "boolean",
        alias: "v",
        default: false,
      },
    },
  }
);

process.on("uncaughtException", function (err) {
  switch (err.errno) {
    case "EACCES":
      console.error(
        "EACCES: run as administrator to use the default ports 443 and 80. " +
          "You can also change port with: `serverino -p 1337`."
      );
      break;
    case "EADDRINUSE":
      console.error(
        "EADDRINUSE: another service on your machine is using " +
          "the current port.\nStop it or change port with:" +
          "`serverino -p 1337`."
      );
      break;
    default:
      console.error("Unexpected error " + err.errno + ":\n\n" + err);
      break;
  }
  process.exit(1);
});

const [root, path] = cli.input;

let _path = path || "";
if (path && !path.startsWith("/")) {
  _path = `/${_path}`;
}

serverino({
  root: root,
  path: _path,
  ...cli.flags,
}).serve();
